### PR TITLE
NO-ISSUE: Cancel stale merge-queue reruns on unit-tests/integration-tests/pre-commit

### PR DESCRIPTION
**Pull request by eliorerz**
#544 scoped full test execution on `merge_group` down to a lightweight compile-check and was correctly rejected -- the owner wants the FULL Unit Tests and Integration Tests suites to keep running on `merge_group` exactly as #418 intended, and live data showed the 180-concurrent-job Enterprise runner ceiling wasn't even close to being hit (~30-40 in use at the time).

**Nothing from #544 is reintroduced here: no compile-check job, no gating the heavy test jobs off of `merge_group`, no change to what runs or when.** This PR only cancels superseded/stale runs.

## The actual problem

None of `unit-tests.yml`, `integration-tests.yml`, or `pre-commit.yaml` cancel a stale run when GitHub's merge queue rebases a PR onto a new ephemeral `gh-readonly-queue` ref -- normal, routine merge-queue behavior that will keep happening regardless of the ref-staleness git-clone incident fixed separately. Every rebase spawns a brand-new full run of these 3 workflows for that PR; the previous run for the now-superseded ref is not cancelled and keeps consuming a runner until it finishes naturally.

Confirmed live: **13 concurrent Unit Tests runs and 10 concurrent Integration Tests runs** against only **4 active merge-queue slots** (`max_entries_to_build: 4`, confirmed via `gh api repos/osac-project/osac/rulesets`) -- consistent with a handful of PRs each stacking up multiple stale, uncancelled reruns from repeated rebases. Also confirmed this repo's real hosted-runner concurrency (~30-40 in use) is nowhere near the plan's 180-job ceiling, so a hard concurrency limit is not the bottleneck here -- the queue stalls because cheap, load-bearing jobs (`label-gate`, `auto-queue`, Slash Command) get starved behind piles of stale heavy runs, not because of a runner cap.

`integration-tests.yml` already had a `concurrency:` block, but its non-PR fallback key was `github.sha` -- the merge-preview commit, which changes on every rebase. So the group key itself changed every rebase and could never collapse a prior run even with `cancel-in-progress` true for `merge_group`. `unit-tests.yml` and `pre-commit.yaml` had no `concurrency:` block at all.

## The fix, and why the key is actually stable

`github.ref_name` for a `merge_group` event is GitHub's ephemeral `gh-readonly-queue/<base>/pr-<number>-<sha>` ref. The `pr-<number>` segment is constant across rebases of the same PR; only the trailing sha changes.

**Verified directly against this repo's own run history** (`gh api repos/osac-project/osac/actions/runs?event=merge_group`), not assumed:

- PR #503 requeued at `gh-readonly-queue/main/pr-503-b2986acb...61` and `.../pr-503-f05965f9...53`, 16 minutes apart.
- PR #307 requeued at `gh-readonly-queue/main/pr-307-18b1c7ed...58` and `.../pr-307-0c86346b...c1`.
- PR #502 requeued at `gh-readonly-queue/main/pr-502-25cf0673...26` and `.../pr-502-ade9ac85...04`.

All three pairs extract to the identical `pr-503`/`pr-307`/`pr-502` via `grep -oE 'pr-[0-9]+'`, despite the trailing sha differing every time -- this is the key that actually collapses repeated rebases.

Workflow-level `concurrency:` blocks are evaluated before any job runs and can't reference a computed value, so this can't be a single top-of-file block (GitHub Actions expressions have no substring-extraction function to pull `pr-<number>` out of the ref inline). Instead:

- **`unit-tests.yml`, `integration-tests.yml`**: the existing `changes` job gets one new step that computes the stable key (`pr-<number>` for `pull_request`, the extracted `pr-<number>` for `merge_group`, `run-<id>` as a no-op fallback for `schedule`/`workflow_dispatch`) and exposes it as a new `concurrency-key` output. Every test-execution job already `needs: changes`, so each gets its own job-level `concurrency:` block referencing `needs.changes.outputs.concurrency-key` -- job-level blocks, unlike workflow-level ones, can reference `needs.*.outputs.*`.
- **`pre-commit.yaml`**: has no `changes` job, so it gets a new, tiny `concurrency-key` job computing the same thing, and `pre-commit` now `needs:` it and carries the same job-level `concurrency:` block.
- `integration-tests.yml`'s old workflow-level block is removed entirely, superseded by the per-job blocks above.

`cancel-in-progress` is `true` for both `pull_request` and `merge_group` on every one of these blocks -- `pull_request` behavior is unchanged in substance (still cancels on new pushes), `merge_group` now actually works.

## Explicitly not changed

- No job is skipped or gated off `merge_group`. Every test-execution job that ran before still runs, on the same triggers, with the same coverage.
- No new compile-check job.
- `pre-commit.yaml`'s actual gitleaks/lint logic is untouched -- only the new upstream `concurrency-key` job and the `needs:`/`concurrency:` addition on `pre-commit` itself.

## Verification

- [x] `actionlint` on all 3 changed files -- clean (also ran actionlint against the whole `.github/workflows/` tree; the only findings are pre-existing, in unrelated files, and not introduced by this PR)
- [x] All 3 files validated as parseable YAML
- [x] Concurrency-key extraction regex verified against 6 real `head_branch` values pulled from this repo's own run history (3 rebase pairs, listed above)
- [ ] A real merge-queue rebase exercises the cancellation end-to-end

## Summary

- **CI:** Added stable per-PR concurrency keys to the unit-test, integration-test, and pre-commit workflows.
- **CI:** Cancel superseded pull-request and merge-group runs with `cancel-in-progress: true`.
- **CI:** Added `pr-<number>` extraction, `run-<run-id>` fallbacks, and restricted permissions for key computation.
- **CI:** Preserved existing triggers, test coverage, and behavior for scheduled, manual, and other events.
- **API, controllers, database, auth, deployment, and documentation:** No changes.

### Backward compatibility

No application or public API behavior changes. CI behavior changes only for overlapping pull-request and merge-group runs. End-to-end cancellation during an actual merge-queue rebase remains unverified.

## Risk classification

**risk:show** — The change affects CI execution and can cancel in-progress workflows, but it does not change application behavior, deployment behavior, public APIs, or test coverage. It was close to **risk:ship** because the scope is limited to workflow configuration, but unverified merge-queue cancellation behavior requires visible review.

### CODE REVIEW

File: .github/workflows/pre-commit.yaml [hunk 1]
```diff
@@ -5,7 +5,52 @@ on:
...
+  # to extract from: for merge_group it's
+  # gh-readonly-queue/<base>/pr-<number>-<sha>, and the pr-<number> segment
+  # stays constant across rebases -- confirmed against this repo's own run
+  # history (gh api .../actions/runs?event=merge_group): PR #503 was
+  # requeued at .../pr-503-b2986acb...61 and .../pr-503-f05965f9...53
+  # sixteen minutes apart, PR #307 similarly at .../pr-307-18b1c7ed...58 and
+  # .../pr-307-0c86346b...c1 -- both pairs extract to the same "pr-<number>"
+  # despite the trailing sha differing every time.
+  concurrency-key:
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (minmzzhang) [reply .github/workflows/pre-commit.yaml]: @eliorerz , might worth to add this one.

**Review by minmzzhang — APPROVED** [review]
/lgtm

may need to address the nit from CR

**Comment by eliorerz** [comment]
Addressed the CodeRabbit actionable comments and the fullsend-ai-review medium/low findings in 415dedbba:

- **Multi-match extraction risk** (CodeRabbit): added \`tail -1\` so the rightmost \`pr-<number>\` match is always used, even if the base branch name itself happened to contain a \`pr-<digits>\`-shaped substring.
- **Unhandled zero-match case** (fullsend-ai-review, Medium): GitHub Actions runs \`run:\` steps with \`set -eo pipefail\`, so a ref that didn't match the expected shape would have made \`grep\` exit 1 and abort the step, cascading to skip every downstream test job via \`needs: changes\`. Added an empty-key fallback to \`run-<run-id>\` (same as the existing schedule/workflow_dispatch fallback) so this degrades to "don't cancel" instead of crashing the workflow.
- **Excess permissions** (CodeRabbit): added \`permissions: {}\` to pre-commit.yaml's new key-computation job -- it does no checkout and makes no API calls, so it needs zero token scope.
- **Naming collision** (fullsend-ai-review, Low): renamed pre-commit.yaml's new job from \`concurrency-key\` to \`compute-concurrency-key\` so it no longer shares a name with the step id used for the same computation in unit-tests.yml/integration-tests.yml's \`changes\` job.

Verified all 3 edge cases (normal ref, multi-match ref, non-matching ref) against the actual bash logic locally, and re-ran \`actionlint\` on all 3 changed files -- clean.

Not changing, with reasoning:
- **protected-path** -- process gate, already satisfied by human review (approved + lgtm).
- **naming-convention** (unit-tests.yml vs integration-tests.yml group-name prefixes) -- cosmetic, no functional effect, not worth the extra diff on an already-reviewed PR.
- **behavioral-change note** on cancel-in-progress now covering merge_group -- that's the intended fix, not a defect; per-PR keying confirmed correct (no cross-PR cancellation).
- **code-organization** (extract the duplicated key-computation script into a composite action) -- valid suggestion, deferring to a follow-up rather than expanding this PR's scope.

**Comment by eliorerz** [comment]
Rebased onto current \`main\` (was 111 commits behind) — clean rebase, no conflicts in the 3 touched workflow files, YAML/actionlint still clean, no other change on main has touched these concurrency blocks in the interim. Confirmed the underlying bug is still live on \`main\` today: \`unit-tests.yml\` and \`pre-commit.yaml\` still have zero concurrency control, and \`integration-tests.yml\` still only has the old workflow-level block keyed on \`github.sha\`.

Addressing CodeRabbit's two review comments:

**Sep 2 (blocking): "use \`github.run_id\` for merge_group events instead"** — this would defeat the entire point of the PR. \`run_id\` is unique per run, so keying on it means concurrent runs never share a group and nothing ever collapses — that's exactly the pileup this PR fixes (13 concurrent Unit Tests runs / 10 Integration Tests runs against 4 merge-queue slots, per the PR description). The whole design is to key on the stable \`pr-<number>\` segment extracted from the ephemeral \`gh-readonly-queue/<base>/pr-<number>-<sha>\` ref specifically so repeated rebases of the *same* PR collapse into one another. Dismissing this one — not applying it, since applying it would reintroduce the bug this PR fixes.

**Aug 27 (informational): stale \`changes\` job could theoretically cancel a newer run instead of an older one** — this is a real, narrow race: the job-level concurrency key is only known once \`changes\` completes, so out-of-order completion of that step could in principle let a stale run's keyed job win the cancellation race against a newer one. The window is small (\`changes\`/\`dorny/paths-filter\` runs in seconds), and critically, even in that worst case you still end up with exactly one run per key instead of an unbounded pileup — so it's a strict improvement over the current state regardless of which run wins the race. Not fixing it here: the fix (e.g. a freshness gate comparing run/queue timestamps before proceeding) adds real complexity for an edge case that's still strictly better than today's guaranteed resource exhaustion. Happy to file a follow-up if it turns out to matter in practice.
